rustdoc: link sibling where possible, even when not One True Path

notriddle opened the pull request:
This is, first of all, a size hack aimed at the `core::arch` modules that share items (x86-64/x86, aarch64/x86). They inline a lot of items from shared modules, and if we can avoid writing `../arm/WHATEVER` all over the aarch64 module, we can shave a few bytes off every link.

Also, clicking a link in the 64 bit module and landing in the 32 bit module makes no sense.

notriddle committed:
rustdoc: link sibling where possible, even when not One True Path

This is, first of all, a size hack aimed at the `core::arch`
modules that share items (x86-64/x86, aarch64/x86). They
inline a lot of items from shared modules, and if we can
avoid writing `../arm/WHATEVER` all over the aarch64 module,
we can shave a few bytes off every link.

Also, clicking a link in the 64 bit module and landing in
the 32 bit module makes no sense.

bors commented:
:umbrella: The latest upstream changes (presumably #124008) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented on src/librustdoc/html/format.rs:
```diff
@@ -725,13 +725,26 @@ pub(crate) fn href_with_root_path(
+            // If this item exists in the current crate, then link to that.
+            (fqp, shortty, {
+                let module_fqp = to_module_fqp(shortty, fqp.as_slice());
+                debug!(?fqp, ?shortty, ?module_fqp);
```
Could you put this `debug!` in the first branch as well please? Or even simpler: after the whole `if let` branches.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit aeeb19735a8ee619202ba56ae9ae9ffa1e0a5e04 with merge bfb0a95051fee4201e00ff20dec4f7609fdbf6ee...
<!-- homu: {"type":"TryBuildStarted","head_sha":"aeeb19735a8ee619202ba56ae9ae9ffa1e0a5e04","merge_sha":"bfb0a95051fee4201e00ff20dec4f7609fdbf6ee"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8741969120/job/23992811331)
Build commit: bfb0a95051fee4201e00ff20dec4f7609fdbf6ee (`bfb0a95051fee4201e00ff20dec4f7609fdbf6ee`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8741969120/job/23992811331"},"merge_sha":"bfb0a95051fee4201e00ff20dec4f7609fdbf6ee"} -->

rust-timer commented:
Queued bfb0a95051fee4201e00ff20dec4f7609fdbf6ee with parent 0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (bfb0a95051fee4201e00ff20dec4f7609fdbf6ee): [comparison URL](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [0.3%, 7.7%] | 4     |
| Regressions ❌ <br /> (secondary)  | 3.1% | [2.2%, 4.0%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [0.3%, 7.7%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean   | range           | count |
|:----------------------------------:|:------:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 171.2% | [15.6%, 319.1%] | 3     |
| Regressions ❌ <br /> (secondary)  | 54.4%  | [5.5%, 103.2%]  | 2     |
| Improvements ✅ <br /> (primary)   | -      | -               | 0     |
| Improvements ✅ <br /> (secondary) | -      | -               | 0     |
| All ❌✅ (primary)                 | 171.2% | [15.6%, 319.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range         | count |
|:----------------------------------:|:-----:|:-------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 10.2% | [3.3%, 17.2%] | 2     |
| Regressions ❌ <br /> (secondary)  | 3.9%  | [3.5%, 4.3%]  | 2     |
| Improvements ✅ <br /> (primary)   | -     | -             | 0     |
| Improvements ✅ <br /> (secondary) | -     | -             | 0     |
| All ❌✅ (primary)                 | 10.2% | [3.3%, 17.2%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0e15f5ee8fb3f9edcb8a35b1cbd9f8d352be54a0&end=bfb0a95051fee4201e00ff20dec4f7609fdbf6ee&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.731s -> 676.763s (0.00%)
**Artifact size**: 316.11 MiB -> 315.33 MiB (-0.25%)

GuillaumeGomez commented:
The impact on `libc` is quite huge (especially on memory).

notriddle committed:
Avoid populating `current_module_linkable_items` with noops

This changes things so that it only gets inlined items added,
instead of populating it with redundant parts.
This is a workaround for `libc`, which has a bunch of items
that get re-exported in one place.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 2868cea81e3db8e19be845f08067154bcc975921 with merge 620d81217e6b871206e01698678714a7d789b48f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2868cea81e3db8e19be845f08067154bcc975921","merge_sha":"620d81217e6b871206e01698678714a7d789b48f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8745228615/job/24002189166)
Build commit: 620d81217e6b871206e01698678714a7d789b48f (`620d81217e6b871206e01698678714a7d789b48f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8745228615/job/24002189166"},"merge_sha":"620d81217e6b871206e01698678714a7d789b48f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8745228615/job/24002189166)
Build commit: 620d81217e6b871206e01698678714a7d789b48f (`620d81217e6b871206e01698678714a7d789b48f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8745228615/job/24002189166"},"merge_sha":"620d81217e6b871206e01698678714a7d789b48f"} -->

rust-timer commented:
Queued 620d81217e6b871206e01698678714a7d789b48f with parent 3412f0174934778dd803081dc7c4b5864309350f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f).
There are currently 4 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~5.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (620d81217e6b871206e01698678714a7d789b48f): [comparison URL](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [0.3%, 7.7%] | 4     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [2.1%, 2.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [0.3%, 7.7%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean   | range            | count |
|:----------------------------------:|:------:|:----------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 128.8% | [0.9%, 318.4%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 101.5% | [101.5%, 101.5%] | 1     |
| Improvements ✅ <br /> (primary)   | -      | -                | 0     |
| Improvements ✅ <br /> (secondary) | -      | -                | 0     |
| All ❌✅ (primary)                 | 128.8% | [0.9%, 318.4%]   | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range         | count |
|:----------------------------------:|:-----:|:-------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 10.6% | [3.5%, 17.7%] | 2     |
| Regressions ❌ <br /> (secondary)  | 4.8%  | [4.8%, 4.8%]  | 1     |
| Improvements ✅ <br /> (primary)   | -     | -             | 0     |
| Improvements ✅ <br /> (secondary) | -     | -             | 0     |
| All ❌✅ (primary)                 | 10.6% | [3.5%, 17.7%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3412f0174934778dd803081dc7c4b5864309350f&end=620d81217e6b871206e01698678714a7d789b48f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 672.221s -> 672.962s (0.11%)
**Artifact size**: 315.28 MiB -> 315.31 MiB (0.01%)
